<a href="https://colab.research.google.com/github/mugambiian/C-coding/blob/main/Football_Score_prediction.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Football Matches Prediction Model

#Question formulation

We will create a model that attempts to predict the outcome of footbal matches between two teams : Team 1 and Team 2 based on five features as below;
1. Home team
2. Away team
3. The type of tournament
4. The rank of the home team
5. The rank of the away team

We shall use regression 

1. Polynomial regression
We shall attempt to predict how many goals the home team scores

Logistic regression

We shall attempt to if the home team wins, looses or draws the game


# Metrics of success

We shall use the Root Mean Squared Error (RMSE) to measure the performace of the model. 

We shall measure the accuracy of the logistic regression model against the accuracy score.



#1.Loading and exploring  the dataset

In [13]:
# we import the libraries we need
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import scipy as sp
from sklearn.preprocessing import StandardScaler, PolynomialFeatures
from sklearn.model_selection import train_test_split, GridSearchCV, KFold, cross_val_score
from sklearn.linear_model import LinearRegression, LogisticRegression
from sklearn.metrics import mean_squared_error, accuracy_score, f1_score, confusion_matrix
import datetime as dt


In [14]:
# We then load our first dataset

Results = pd.read_csv("/content/results.csv")
Results

date         home_team  ...   country  neutral
0      1872-11-30          Scotland  ...  Scotland    False
1      1873-03-08           England  ...   England    False
2      1874-03-07          Scotland  ...  Scotland    False
3      1875-03-06           England  ...   England    False
4      1876-03-04          Scotland  ...  Scotland    False
...           ...               ...  ...       ...      ...
40834  2019-07-18    American Samoa  ...     Samoa     True
40835  2019-07-18              Fiji  ...     Samoa     True
40836  2019-07-19           Senegal  ...     Egypt     True
40837  2019-07-19        Tajikistan  ...     India     True
40838  2019-07-20  Papua New Guinea  ...     Samoa     True

[40839 rows x 9 columns]

In [15]:
# We then load our first dataset

Results = pd.read_csv("/content/results.csv")
Results

date         home_team  ...   country  neutral
0      1872-11-30          Scotland  ...  Scotland    False
1      1873-03-08           England  ...   England    False
2      1874-03-07          Scotland  ...  Scotland    False
3      1875-03-06           England  ...   England    False
4      1876-03-04          Scotland  ...  Scotland    False
...           ...               ...  ...       ...      ...
40834  2019-07-18    American Samoa  ...     Samoa     True
40835  2019-07-18              Fiji  ...     Samoa     True
40836  2019-07-19           Senegal  ...     Egypt     True
40837  2019-07-19        Tajikistan  ...     India     True
40838  2019-07-20  Papua New Guinea  ...     Samoa     True

[40839 rows x 9 columns]

The object data type will beed to be changed to something more appropriate for manipulation

In [16]:
#we load our second dataset
Rankings = pd.read_csv('/content/fifa_ranking.csv')
Rankings

rank country_full  ... confederation   rank_date
0         1      Germany  ...          UEFA  1993-08-08
1         2        Italy  ...          UEFA  1993-08-08
2         3  Switzerland  ...          UEFA  1993-08-08
3         4       Sweden  ...          UEFA  1993-08-08
4         5    Argentina  ...      CONMEBOL  1993-08-08
...     ...          ...  ...           ...         ...
57788   206     Anguilla  ...      CONCACAF  2018-06-07
57789   206      Bahamas  ...      CONCACAF  2018-06-07
57790   206      Eritrea  ...           CAF  2018-06-07
57791   206      Somalia  ...           CAF  2018-06-07
57792   206        Tonga  ...           OFC  2018-06-07

[57793 rows x 16 columns]

In [17]:
#  We look at the data types for the various columns
Rankings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 57793 entries, 0 to 57792
Data columns (total 16 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   rank                     57793 non-null  int64  
 1   country_full             57793 non-null  object 
 2   country_abrv             57793 non-null  object 
 3   total_points             57793 non-null  float64
 4   previous_points          57793 non-null  int64  
 5   rank_change              57793 non-null  int64  
 6   cur_year_avg             57793 non-null  float64
 7   cur_year_avg_weighted    57793 non-null  float64
 8   last_year_avg            57793 non-null  float64
 9   last_year_avg_weighted   57793 non-null  float64
 10  two_year_ago_avg         57793 non-null  float64
 11  two_year_ago_weighted    57793 non-null  float64
 12  three_year_ago_avg       57793 non-null  float64
 13  three_year_ago_weighted  57793 non-null  float64
 14  confederation         

As before the object data type will need to be edited to a more suitable type for manipulation



# 2.Data cleanup

In [18]:
# We take a closer look at our dataset

Results.describe()

home_score    away_score
count  40839.000000  40839.000000
mean       1.745709      1.188105
std        1.749145      1.405120
min        0.000000      0.000000
25%        1.000000      0.000000
50%        1.000000      1.000000
75%        2.000000      2.000000
max       31.000000     21.000000

In [19]:
# we check for null values

Results.isnull().sum()

date          0
home_team     0
away_team     0
home_score    0
away_score    0
tournament    0
city          0
country       0
neutral       0
dtype: int64

We don't have any null values


In [20]:
#We check for duplicated values
Results.duplicated().any()

False

We don't have any duplicates as well



In [21]:
# We rename the columns for consistency
Results = Results.rename(columns = {
    'date' : 'Date',
    'home_team ' : 'Home_team',
    'home_score' : 'Home_Score',
    'away_score' : 'Away_Score',
    'tournament' : 'Tournament',
    'city' : 'City',
    'country' : 'Country',
    'neutral' : 'Neutral'

})
Results

Date         home_team  ...   Country  Neutral
0      1872-11-30          Scotland  ...  Scotland    False
1      1873-03-08           England  ...   England    False
2      1874-03-07          Scotland  ...  Scotland    False
3      1875-03-06           England  ...   England    False
4      1876-03-04          Scotland  ...  Scotland    False
...           ...               ...  ...       ...      ...
40834  2019-07-18    American Samoa  ...     Samoa     True
40835  2019-07-18              Fiji  ...     Samoa     True
40836  2019-07-19           Senegal  ...     Egypt     True
40837  2019-07-19        Tajikistan  ...     India     True
40838  2019-07-20  Papua New Guinea  ...     Samoa     True

[40839 rows x 9 columns]

In [22]:
#we take a closer look at our second dataset
Rankings.describe()


rank  total_points  ...  three_year_ago_avg  three_year_ago_weighted
count  57793.000000  57793.000000  ...        57793.000000             57793.000000
mean     101.628086    122.068637  ...           59.173916                11.834811
std       58.618424    260.426863  ...          135.533343                27.106675
min        1.000000      0.000000  ...            0.000000                 0.000000
25%       51.000000      0.000000  ...            0.000000                 0.000000
50%      101.000000      0.000000  ...            0.000000                 0.000000
75%      152.000000     92.790000  ...           21.250000                 4.250000
max      209.000000   1775.030000  ...         1200.770000               240.150000

[8 rows x 12 columns]

In [23]:
# we check for null values

Results.isnull().sum()

Date          0
home_team     0
away_team     0
Home_Score    0
Away_Score    0
Tournament    0
City          0
Country       0
Neutral       0
dtype: int64

  We don't have any null values
  

In [24]:
#We rename the columns for consistency
Rankings = Rankings.rename(columns = {
    'country_full' : 'Country_Full',
    'country_abrv' : 'Country_Abrv',
    'total_points' : 'Total_Points',
    'rank_change' : 'Rank_Change',
    'cur_year_avg' : 'Cur_year_Average',
    'cur_year_avg_weighted' : 'Cur_Year_Avg_Weighted',
    'last_year_avg' : 'Last_Year_Avg',
    'last_year_avg_weighted' : 'Last_Year_Avg_Weighted',
    'two_year_ago_avg' : 'Level_of_Education',
    'two_year_ago_weighted' : 'Two_Year_Ago_Weighted',
    'three_year_ago_avg' : 'Three_Year_Ago_Avg',
    'three_year_ago_weighted' : 'Three_Year_Ago_Weighted',
     'confederation' : 'Confederation',
    'rank_date' : 'Date'

})
Rankings

rank Country_Full  ... Confederation        Date
0         1      Germany  ...          UEFA  1993-08-08
1         2        Italy  ...          UEFA  1993-08-08
2         3  Switzerland  ...          UEFA  1993-08-08
3         4       Sweden  ...          UEFA  1993-08-08
4         5    Argentina  ...      CONMEBOL  1993-08-08
...     ...          ...  ...           ...         ...
57788   206     Anguilla  ...      CONCACAF  2018-06-07
57789   206      Bahamas  ...      CONCACAF  2018-06-07
57790   206      Eritrea  ...           CAF  2018-06-07
57791   206      Somalia  ...           CAF  2018-06-07
57792   206        Tonga  ...           OFC  2018-06-07

[57793 rows x 16 columns]

In [25]:
Rankings.Date = Rankings.Date.apply(lambda x: x.replace('/', '-'))
Results.Date = Results.Date.apply(lambda x: x.replace('/', '-'))


In [26]:
#We change our date to date time 
Rankings.Date = pd.to_datetime(Rankings.Date)
Results.Date = pd.to_datetime(Results.Date)

In [27]:
#We need to split the date into year and month
Rankings['year'] = Rankings.Date.dt.year
Results['year'] = Results.Date.dt.year

Rankings['month'] = Rankings.Date.dt.month
Results['month'] = Rankings.Date.dt.month




In [28]:
#We confirm that the date columns have changed to datetime objects
Rankings.info()
Results.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 57793 entries, 0 to 57792
Data columns (total 18 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   rank                     57793 non-null  int64         
 1   Country_Full             57793 non-null  object        
 2   Country_Abrv             57793 non-null  object        
 3   Total_Points             57793 non-null  float64       
 4   previous_points          57793 non-null  int64         
 5   Rank_Change              57793 non-null  int64         
 6   Cur_year_Average         57793 non-null  float64       
 7   Cur_Year_Avg_Weighted    57793 non-null  float64       
 8   Last_Year_Avg            57793 non-null  float64       
 9   Last_Year_Avg_Weighted   57793 non-null  float64       
 10  Level_of_Education       57793 non-null  float64       
 11  Two_Year_Ago_Weighted    57793 non-null  float64       
 12  Three_Year_Ago_Avg       57793 n

In [29]:
#We combine the two datasets to begin our analysis
Combined_Data = pd.merge(Results, Rankings, how = 'left', left_on = ['year', 'month'], right_on = ['year', 'month'])
Combined_Data

Date_x         home_team  ... Confederation  Date_y
0       1872-11-30          Scotland  ...           NaN     NaT
1       1873-03-08           England  ...           NaN     NaT
2       1874-03-07          Scotland  ...           NaN     NaT
3       1875-03-06           England  ...           NaN     NaT
4       1876-03-04          Scotland  ...           NaN     NaT
...            ...               ...  ...           ...     ...
4478938 2019-07-18    American Samoa  ...           NaN     NaT
4478939 2019-07-18              Fiji  ...           NaN     NaT
4478940 2019-07-19           Senegal  ...           NaN     NaT
4478941 2019-07-19        Tajikistan  ...           NaN     NaT
4478942 2019-07-20  Papua New Guinea  ...           NaN     NaT

[4478943 rows x 27 columns]

In [30]:
#We preview our datasets
Combined_Data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4478943 entries, 0 to 4478942
Data columns (total 27 columns):
 #   Column                   Dtype         
---  ------                   -----         
 0   Date_x                   datetime64[ns]
 1   home_team                object        
 2   away_team                object        
 3   Home_Score               int64         
 4   Away_Score               int64         
 5   Tournament               object        
 6   City                     object        
 7   Country                  object        
 8   Neutral                  bool          
 9   year                     int64         
 10  month                    int64         
 11  rank                     float64       
 12  Country_Full             object        
 13  Country_Abrv             object        
 14  Total_Points             float64       
 15  previous_points          float64       
 16  Rank_Change              float64       
 17  Cur_year_Average         fl

In [31]:
Combined_Data.describe

<bound method NDFrame.describe of             Date_x         home_team  ... Confederation  Date_y
0       1872-11-30          Scotland  ...           NaN     NaT
1       1873-03-08           England  ...           NaN     NaT
2       1874-03-07          Scotland  ...           NaN     NaT
3       1875-03-06           England  ...           NaN     NaT
4       1876-03-04          Scotland  ...           NaN     NaT
...            ...               ...  ...           ...     ...
4478938 2019-07-18    American Samoa  ...           NaN     NaT
4478939 2019-07-18              Fiji  ...           NaN     NaT
4478940 2019-07-19           Senegal  ...           NaN     NaT
4478941 2019-07-19        Tajikistan  ...           NaN     NaT
4478942 2019-07-20  Papua New Guinea  ...           NaN     NaT

[4478943 rows x 27 columns]>

In [32]:
Combined_Data


Date_x         home_team  ... Confederation  Date_y
0       1872-11-30          Scotland  ...           NaN     NaT
1       1873-03-08           England  ...           NaN     NaT
2       1874-03-07          Scotland  ...           NaN     NaT
3       1875-03-06           England  ...           NaN     NaT
4       1876-03-04          Scotland  ...           NaN     NaT
...            ...               ...  ...           ...     ...
4478938 2019-07-18    American Samoa  ...           NaN     NaT
4478939 2019-07-18              Fiji  ...           NaN     NaT
4478940 2019-07-19           Senegal  ...           NaN     NaT
4478941 2019-07-19        Tajikistan  ...           NaN     NaT
4478942 2019-07-20  Papua New Guinea  ...           NaN     NaT

[4478943 rows x 27 columns]

In [33]:
# We select the Home Team details
Home_Team_Combined = Combined_Data[Combined_Data.home_team == Combined_Data.Country_Full]
Home_Team_Combined.head()



Date_x   home_team  ... Confederation     Date_y
18192 1994-01-02    Barbados  ...      CONCACAF 1994-04-19
18291 1994-01-02       Ghana  ...           CAF 1994-04-19
18479 1994-01-05        Mali  ...           CAF 1994-04-19
18747 1994-01-09  Mauritania  ...           CAF 1994-04-19
18838 1994-01-11    Thailand  ...           AFC 1994-04-19

[5 rows x 27 columns]

In [34]:
# We select the Away Team details
Away_Team_Combined = Combined_Data[Combined_Data.away_team == Combined_Data.Country_Full]
Away_Team_Combined.head()

Date_x   home_team  ... Confederation     Date_y
18224 1994-01-02    Barbados  ...      CONCACAF 1994-04-19
18286 1994-01-02       Ghana  ...           CAF 1994-04-19
18532 1994-01-05        Mali  ...           CAF 1994-04-19
18648 1994-01-09  Mauritania  ...           CAF 1994-04-19
18774 1994-01-11    Thailand  ...           CAF 1994-04-19

[5 rows x 27 columns]

In [35]:
#we dont need all the columns so we select a subset of only what we need

Home_Team_Combined.columns
Away_Team_Combined.columns

Index(['Date_x', 'home_team', 'away_team', 'Home_Score', 'Away_Score',
       'Tournament', 'City', 'Country', 'Neutral', 'year', 'month', 'rank',
       'Country_Full', 'Country_Abrv', 'Total_Points', 'previous_points',
       'Rank_Change', 'Cur_year_Average', 'Cur_Year_Avg_Weighted',
       'Last_Year_Avg', 'Last_Year_Avg_Weighted', 'Level_of_Education',
       'Two_Year_Ago_Weighted', 'Three_Year_Ago_Avg',
       'Three_Year_Ago_Weighted', 'Confederation', 'Date_y'],
      dtype='object')

In [36]:
# Selecting only the relevant columns in the away_combined dataframe
# Creating a list of relevant columns
#
Combined_Data_Cleaned = list(Away_Team_Combined.columns[9:-2])
Combined_Data_Cleaned.extend(['home_team', 'away_team'])

Away_Team_Combined = Away_Team_Combined[Combined_Data_Cleaned]
Away_Team_Combined

year  month   rank  ... Three_Year_Ago_Weighted    home_team     away_team
18224    1994      4  131.0  ...                    0.00     Barbados       Grenada
18286    1994      4   24.0  ...                    0.00        Ghana         Egypt
18532    1994      4  101.0  ...                    0.00         Mali  Burkina Faso
18648    1994      4   48.0  ...                    0.00   Mauritania          Mali
18774    1994      4    5.0  ...                    0.00     Thailand       Nigeria
...       ...    ...    ...  ...                     ...          ...           ...
4477065  2018      6   55.0  ...                   41.86  South Korea        Panama
4477407  2018      6  186.0  ...                   25.50         Laos      Mongolia
4477652  2018      6    9.0  ...                   97.51       Mexico         Chile
4477914  2018      6   60.0  ...                   73.95         Oman       Ecuador
4478093  2018      6   28.0  ...                  147.46       Sweden      Slovakia

[19267 rows x 18 columns]

In [37]:
# We need to combine the Home Team and Away Team data frames to then proceed to answer our reserach questions
Football_Teams_Data = pd.merge(Home_Team_Combined, Away_Team_Combined, how = 'left' , left_on =['year', 'month', 'home_team', 'away_team'],\
                    right_on = ['year', 'month', 'home_team', 'away_team'])
Football_Teams_Data

Date_x   home_team  ... Three_Year_Ago_Avg_y  Three_Year_Ago_Weighted_y
0     1994-01-02    Barbados  ...                 0.00                       0.00
1     1994-01-02    Barbados  ...                 0.00                       0.00
2     1994-01-02       Ghana  ...                 0.00                       0.00
3     1994-01-05        Mali  ...                 0.00                       0.00
4     1994-01-09  Mauritania  ...                 0.00                       0.00
...          ...         ...  ...                  ...                        ...
20573 2018-10-16        Laos  ...               127.50                      25.50
20574 2018-10-16    Malaysia  ...                  NaN                        NaN
20575 2018-10-16      Mexico  ...               487.56                      97.51
20576 2018-10-16        Oman  ...               369.77                      73.95
20577 2018-10-16      Sweden  ...               737.30                     147.46

[20578 rows x 41 columns]

In [38]:
# We drop columns that we will not need for our analysis
drop_cols = ['Country_Full_x', 'Country_Abrv_x', 'Date_y', 'Country_Full_y',	'Country_Abrv_y']
Football_Teams_Data.drop(drop_cols, axis = 1, inplace = True)


In [39]:
Football_Teams_Data.columns


Index(['Date_x', 'home_team', 'away_team', 'Home_Score', 'Away_Score',
       'Tournament', 'City', 'Country', 'Neutral', 'year', 'month', 'rank_x',
       'Total_Points_x', 'previous_points_x', 'Rank_Change_x',
       'Cur_year_Average_x', 'Cur_Year_Avg_Weighted_x', 'Last_Year_Avg_x',
       'Last_Year_Avg_Weighted_x', 'Level_of_Education_x',
       'Two_Year_Ago_Weighted_x', 'Three_Year_Ago_Avg_x',
       'Three_Year_Ago_Weighted_x', 'Confederation', 'rank_y',
       'Total_Points_y', 'previous_points_y', 'Rank_Change_y',
       'Cur_year_Average_y', 'Cur_Year_Avg_Weighted_y', 'Last_Year_Avg_y',
       'Last_Year_Avg_Weighted_y', 'Level_of_Education_y',
       'Two_Year_Ago_Weighted_y', 'Three_Year_Ago_Avg_y',
       'Three_Year_Ago_Weighted_y'],
      dtype='object')

In [40]:
#We rename our Columns for consitency
Football_Teams_Data = Football_Teams_Data.rename(columns={
    'rank_x' : 'Rank_Home',
    'Total_Points_x' : 'Total_Points_Home',
    'previous_points_x' :'Previous_Points_Home', 
     'Rank_Change_x':'Rank_Change_Home',
     'Cur_year_Average_x': 'Curv_year_Average_Home', 
     'Cur_Year_Avg_Weighted_x' : 'Curv_Year_Avg_Weighted_Home', 
     'Last_Year_Avg_x' : 'Last_Year_Avg_Home',
     'Last_Year_Avg_Weighted_x' : 'Last_Year_Avg_Weighted_Home', 
     'Level_of_Education_x' : 'Level_of_Education_Home',
     'Two_Year_Ago_Weighted_x' : 'Two_Year_Ago_Weighted_Home', 
     'Three_Year_Ago_Avg_x' : 'Three_Year_Ago_Avg_Home',
     'Three_Year_Ago_Weighted_x' : 'Three_Year_Ago_Weighted_Home', 
     'rank_y': 'Rank_Away',
     'Total_Points_y': 'Total_Points_Away', 
     'previous_points_y' : 'Previous_Points_Away', 
     'Rank_Change_y' : 'Rank_Change_Away',
     'Cur_year_Average_y': 'Cur_year_Average_Away', 
     'Cur_Year_Avg_Weighted_y':'Cur_Year_Avg_Weighted_Away', 
     'Last_Year_Avg_y':'Last_Year_Avg_Away',
     'Last_Year_Avg_Weighted_y': 'Last_Year_Avg_Weighted_Away',
     'Level_of_Education_y':'Level_of_Education_Away',
     'Two_Year_Ago_Weighted_y':'Two_Year_Ago_Weighted_Away', 
     'Three_Year_Ago_Avg_y':'Three_Year_Ago_Avg_Away',
     'Three_Year_Ago_Weighted_y' :'Three_Year_Ago_Weighted_Away'
})
Football_Teams_Data.columns


Index(['Date_x', 'home_team', 'away_team', 'Home_Score', 'Away_Score',
       'Tournament', 'City', 'Country', 'Neutral', 'year', 'month',
       'Rank_Home', 'Total_Points_Home', 'Previous_Points_Home',
       'Rank_Change_Home', 'Curv_year_Average_Home',
       'Curv_Year_Avg_Weighted_Home', 'Last_Year_Avg_Home',
       'Last_Year_Avg_Weighted_Home', 'Level_of_Education_Home',
       'Two_Year_Ago_Weighted_Home', 'Three_Year_Ago_Avg_Home',
       'Three_Year_Ago_Weighted_Home', 'Confederation', 'Rank_Away',
       'Total_Points_Away', 'Previous_Points_Away', 'Rank_Change_Away',
       'Cur_year_Average_Away', 'Cur_Year_Avg_Weighted_Away',
       'Last_Year_Avg_Away', 'Last_Year_Avg_Weighted_Away',
       'Level_of_Education_Away', 'Two_Year_Ago_Weighted_Away',
       'Three_Year_Ago_Avg_Away', 'Three_Year_Ago_Weighted_Away'],
      dtype='object')

In [41]:
#We check for missing values
Football_Teams_Data.isnull().sum()

Date_x                             0
home_team                          0
away_team                          0
Home_Score                         0
Away_Score                         0
Tournament                         0
City                               0
Country                            0
Neutral                            0
year                               0
month                              0
Rank_Home                          0
Total_Points_Home                  0
Previous_Points_Home               0
Rank_Change_Home                   0
Curv_year_Average_Home             0
Curv_Year_Avg_Weighted_Home        0
Last_Year_Avg_Home                 0
Last_Year_Avg_Weighted_Home        0
Level_of_Education_Home            0
Two_Year_Ago_Weighted_Home         0
Three_Year_Ago_Avg_Home            0
Three_Year_Ago_Weighted_Home       0
Confederation                      0
Rank_Away                       1811
Total_Points_Away               1811
Previous_Points_Away            1811
R

We have a large number of missing values that we will drop

In [42]:
#we drop the missing values
Football_Teams_Data.dropna(inplace=True)

In [43]:
#we confirm that our changes have been succesful
Football_Teams_Data.isnull().sum()

Date_x                          0
home_team                       0
away_team                       0
Home_Score                      0
Away_Score                      0
Tournament                      0
City                            0
Country                         0
Neutral                         0
year                            0
month                           0
Rank_Home                       0
Total_Points_Home               0
Previous_Points_Home            0
Rank_Change_Home                0
Curv_year_Average_Home          0
Curv_Year_Avg_Weighted_Home     0
Last_Year_Avg_Home              0
Last_Year_Avg_Weighted_Home     0
Level_of_Education_Home         0
Two_Year_Ago_Weighted_Home      0
Three_Year_Ago_Avg_Home         0
Three_Year_Ago_Weighted_Home    0
Confederation                   0
Rank_Away                       0
Total_Points_Away               0
Previous_Points_Away            0
Rank_Change_Away                0
Cur_year_Average_Away           0
Cur_Year_Avg_W

In [44]:
#we check the data types in the merged data set
Football_Teams_Data.dtypes


Date_x                          datetime64[ns]
home_team                               object
away_team                               object
Home_Score                               int64
Away_Score                               int64
Tournament                              object
City                                    object
Country                                 object
Neutral                                   bool
year                                     int64
month                                    int64
Rank_Home                              float64
Total_Points_Home                      float64
Previous_Points_Home                   float64
Rank_Change_Home                       float64
Curv_year_Average_Home                 float64
Curv_Year_Avg_Weighted_Home            float64
Last_Year_Avg_Home                     float64
Last_Year_Avg_Weighted_Home            float64
Level_of_Education_Home                float64
Two_Year_Ago_Weighted_Home             float64
Three_Year_Ag

In [45]:
#We check for duplicated values
Football_Teams_Data.duplicated().any()

True

In [46]:
# we drop the duplicates
Football_Teams_Data.drop_duplicates(keep = 'first',inplace=True)

In [47]:
#We confirm that the duplicate values have been dropped
Football_Teams_Data.duplicated().any()

False

In [48]:
Football_Teams_Data_Backup = Football_Teams_Data
Football_Teams_Data_Backup

Date_x  ... Three_Year_Ago_Weighted_Away
0     1994-01-02  ...                         0.00
2     1994-01-02  ...                         0.00
3     1994-01-05  ...                         0.00
4     1994-01-09  ...                         0.00
5     1994-01-11  ...                         0.00
...          ...  ...                          ...
20572 2018-10-16  ...                       119.85
20573 2018-10-16  ...                        25.50
20575 2018-10-16  ...                        97.51
20576 2018-10-16  ...                        73.95
20577 2018-10-16  ...                       147.46

[17840 rows x 36 columns]

In [49]:
Football_Teams_Data_Backup['Result'] = Football_Teams_Data_Backup.Home_Score - Football_Teams_Data_Backup.Away_Score
Football_Teams_Data_Backup.Result

0        0
2        1
3        0
4       -2
5        0
        ..
20572    1
20573   -3
20575   -1
20576    0
20577    0
Name: Result, Length: 17840, dtype: int64

In [50]:
def  Result(goals):
  if goals > 0:
    return 'Win'
  elif goals < 0:
    return 'Lose'
  else:
    return 'Draw'

#We assign our resuls to our dataframe

Football_Teams_Data_Backup['Result'] = Football_Teams_Data_Backup['Result'].apply(lambda goals: Result(goals))
Football_Teams_Data_Backup



Date_x   home_team  ... Three_Year_Ago_Weighted_Away  Result
0     1994-01-02    Barbados  ...                         0.00    Draw
2     1994-01-02       Ghana  ...                         0.00     Win
3     1994-01-05        Mali  ...                         0.00    Draw
4     1994-01-09  Mauritania  ...                         0.00    Lose
5     1994-01-11    Thailand  ...                         0.00    Draw
...          ...         ...  ...                          ...     ...
20572 2018-10-16       Japan  ...                       119.85     Win
20573 2018-10-16        Laos  ...                        25.50    Lose
20575 2018-10-16      Mexico  ...                        97.51    Lose
20576 2018-10-16        Oman  ...                        73.95    Draw
20577 2018-10-16      Sweden  ...                       147.46    Draw

[17840 rows x 37 columns]

In [51]:
Football_Teams_Data_Backup 

Date_x   home_team  ... Three_Year_Ago_Weighted_Away  Result
0     1994-01-02    Barbados  ...                         0.00    Draw
2     1994-01-02       Ghana  ...                         0.00     Win
3     1994-01-05        Mali  ...                         0.00    Draw
4     1994-01-09  Mauritania  ...                         0.00    Lose
5     1994-01-11    Thailand  ...                         0.00    Draw
...          ...         ...  ...                          ...     ...
20572 2018-10-16       Japan  ...                       119.85     Win
20573 2018-10-16        Laos  ...                        25.50    Lose
20575 2018-10-16      Mexico  ...                        97.51    Lose
20576 2018-10-16        Oman  ...                        73.95    Draw
20577 2018-10-16      Sweden  ...                       147.46    Draw

[17840 rows x 37 columns]

#Polynomial Regression

In [5]:
#we will use a smaller subset of the Columns available and pick those
Poly_Data = Football_Teams_Data_Backup[['home_team',	'away_team',	'Home_Score',	'Away_Score',	'Tournament', 'Rank_Home', 'Rank_Away']]
Poly_Data

NameError: ignored

# We create our polynomial model

In [4]:

X = Poly_Data.iloc[:, [0, 1, 3, 4, 5, 6]]
y = Poly_Data.Home_Score


X = pd.get_dummies(X, drop_first=True)

# We then split the dataset into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25, random_state = 101)

# We then standardise the X_train and the X_test to the same scale
#
SC = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

# We fit the polynomial features to the X_train and X_test and train the model
#
Poly_Features = PolynomialFeatures(degree = 1)
X_train = poly_features.fit_transform(X_train)
X_test = poly_features.fit_transform(X_test)

Regressor = LinearRegression()
Regressor.fit(X_train, y_train)


y_pred = regressor.predict(X_test)

# We measure the accuracy of our model
#
print(np.sqrt(mean_squared_error(y_test, y_pred)))

NameError: ignored

We have a high RSME Score which is bad  as a lowe sxcore indicates a better probability of predicting the score

In [63]:
# We do hyper parameter tuning to find the optimal parameters
from sklearn.linear_model import LinearRegression
LinearRegression()

params = {'normalize': [True, False],
         'fit_intercept': [True, False]}
         
kfold = KFold(n_splits = 10)

Grid_Search = GridSearchCV(estimator=regressor, param_grid = params, cv = kfold, scoring = 'neg_mean_squared_error')

Grid_Search_results = Grid_Search.fit(X, y)

Grid_Search.best_score_
Grid_Search.best_params_




KeyboardInterrupt: ignored

#Logistic Regression


In [3]:
#we will use a smaller subset of the Columns available and pick those

Logistic_Data = Football_Teams_Data_Backup[['home_team', 'away_team', 'Home_Score', 'Away_Score', 'Tournament', 'year', 'Rank_Home', 'Rank_Away', 'Result']]

Logistic_Data



NameError: ignored

In [59]:
#We split the data into features and the target variable
X = Logistic_Data.drop('Result', axis = 1)
y = Logistic_Data.Result

X = pd.get_dummies(X, drop_first=True)

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25, random_state = 101)

Logistic = LogisticRegression()
Logistic.fit(X_train, y_train)

y_pred = logistic.predict(X_test)

# We measure the accuracy of the model

print(f'The accuracy of the model is {accuracy_score(y_test, y_pred)}')
print(f'The f1 score of the model is {f1_score(y_test, y_pred, average = "macro")}')



The accuracy of the model is 0.8926008968609865
The f1 score of the model is 0.8712933805879883


/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


We have an accuracy score of 89% and an f1 score of 0.87. A higher accuracy and f1 score is good and since it isnt 100% , we havent overfitted as well